Multiple clients via features

marioortizmanero opened the pull request:
This is a second attempt for fixing #112. The original idea at #120 was to have a `blocking` module that would call the async implementations with a blocking runtime in order to avoid code duplication. This turned up to be a bad idea, because it wasn't as easy and flexible as I thought. This new approach uses the [`maybe-async`](https://docs.rs/maybe-async/0.2.0/maybe_async/) crate, which makes it relatively easy to use multiple HTTP clients in the same crate with features. Instead of having a `blocking` module, the user may configure what client is used like so: `rspotify = { version = "...", default_features = false, features = "client-ureq" }` (`client-reqwest` is still the default).

## Advantages of this approach:
* Much cleaner implementation for the "blocking module"
* More flexible than a switch between sync and async code, since multiple clients for each of these can be configured. For now it's only two. This flexibility also makes rspotify faster and less bloated, since the blocking module now uses `ureq` instead of `reqwest`, which is much more minimal.
* Just as fast to compile: you'd think that the original approach would be faster to compile because it doesn't *need* a macro to work. But in the end the macro is necessary to make development less of a pain, so both solutions use macros. I haven't tested this but I'm sure the difference isn't noticeable.
* The endpoint function signatures aren't limited: the original approach required a `macro_rules!` macro. These are quick and easy to implement, but supporting generic parameters and other compex stuff can get out of hands pretty quickly. This avoids any workarounds for that.
* The `derive_deref` dependency isn't necessary.

## Disadvantages of this approach:
* Only one client may be used at once
* New dependency: [`maybe-async`](https://docs.rs/maybe-async/0.2.0/maybe_async/). It's just a procedural macro so no performance impact whatsoever. This also requires `async_trait`.

## Required changes and progress
- [x] Async client with `reqwest`
- [x] Blocking client with `ureq`
- [x] Rewrite OAuth
- [x] Update documentation
- [x] Update tests, make them available for all clients
- [x] Update basic examples
- [x] Update the `webapp` example
- [x] Open an issue at `maybe_async` about the parenthesis warnings when using `ureq`: https://github.com/fMeow/maybe-async-rs/issues/1

Read the CHANGELOG.md additions for more changes (most of them are from #127). I had to rewrite the `fetch_access_token` logic so that it can use the HTTP client in `Spotify`, and I ended up improving lots of other stuff I saw. (this is still very WIP, along with the first two points of this section).

I've left lots of `TODO`s that can be discussed after this is merged. I'd like to finish this ASAP so that other changes can be made without merge conflicts and without having to apply them for the blocking module as well.

Also see the examples, some of them have been modified to show how the API works now.

* Closes #109
* Closes #112 
* Closes #116
* Closes #131

marioortizmanero commented:
The implementation is almost finished... What I need to figure out is why I'm getting error 400: Bad Request from the endpoints...

marioortizmanero commented:
The requests are working now! All that's left is fixing the tests and the docs.

ramsayleung commented:
I am still looking at the updated code, I still have question about how `may_async` switch between `async` version and `sync` version, how does it work?

ramsayleung commented on src/client.rs:
```diff
@@ -100,140 +109,417 @@ pub enum ApiError {
+    /// The Spotify API prefix, `https://api.spotify.com/v1/` by default.
+    #[builder(
+        setter(into),
+        default = "String::from(\"https://api.spotify.com/v1/\")"
```
how about using `r##r` operator instead of escaped string?

marioortizmanero replied:
Good idea, I'll fix that.

marioortizmanero commented:
> I am still looking at the updated code, I still have question about how may_async switch between async version and sync version, how does it work?

`maybe_async` is triggered inside `Cargo.toml`. It will be asynchronous by default, but if the `maybe_async/is_sync` feature is enabled (which it is in the case of `client-ureq`), the methods will be synchronous. It's exactly what I was planning on implementing at the very beginning to switch between async and sync, but it's already a thing, and pretty well done.

> I have run the album example, everything is okay, perhaps I missed the failed one, which endpoint failed?

The `album` endpoint is working, but some other endpoints aren't. I think it's because the parameters aren't passed correctly (`album` doesn't need parameters). Run the tests with `cargo test --features=cli,env-file`

marioortizmanero commented:
You're completely right. Thanks for your insight :) Here's the final diff with your suggestions in the commit https://github.com/ramsayleung/rspotify/pull/129/commits/6348d01baa8e3070cdea1f66190ae9ee496a4c6e. I wasn't sure what to do for a `Form` with `GET`, so I just added an assert in that case. Or should it just return the same `request` like the `Json` case does? It's a bit pointless because `BaseClient` only has a `post_form` method, but I would make it as robust as possible to make it more future-proof.

```diff

-        let mut request = self.client.request(method, &url).headers(headers);
+        let mut request = self.client.request(method.clone(), &url).headers(headers);

         log::info!("Making request {:?} with payload {:?}", request, payload);

         if let Some(payload) = payload {
             request = match payload {
-                Content::Json(value) => request.json(value),
-                Content::Form(value) => request.form(value),
+                Content::Json(payload) => match method {
+                    Method::GET => request.query(payload),
+                    Method::POST | Method::PUT | Method::DELETE => request.json(payload),
+                    // Method: Options, Head, Trace haven't implemented in `rspotify` yet, just leave it alone.
+                    _ => request,
+                }
+                Content::Form(payload) => {
+                    // `Request::form` won't work for `GET` requests
+                    assert!(method != Method::GET);
+                    request.form(payload)
+                }
             };
         }
```

>  And we can't simply use ContentType to match form and json function, just try to use MethodType instead, perhaps you could take a look at this PR for more details: 

I don't really understand what you meant with this. I need to differentiate between JSON and Form requests because the `internal_call` method only worked for JSON requests. For example [here you had to use a custom reqwest call](https://github.com/ramsayleung/rspotify/blob/367735611b451212c5d695ade5c1cfde4e849c02/src/oauth2.rs#L472), which is impossible when we may have multiple clients.

I've fixed the remaining errors. I was getting a `Length Required` one because POST, PUT and DELETE methods can't have empty payloads. I'm still trying to figure out why I get some `TrySendError { kind: Disconnected }` errors, and fixing the CI.

ramsayleung commented:
>I don't really understand what you meant with this. I need to differentiate between JSON and Form requests because the internal_call method only worked for JSON requests. For example here you had to use a custom reqwest call, which is impossible when we may have multiple clients.

Yep, I just forget we are mirgating to multiple clients, perhaps we should figure out other solution.

> I've fixed the remaining errors. I was getting a Length Required one because POST, PUT and DELETE methods can't have empty payloads. I'm still trying to figure out why I get some TrySendError { kind: Disconnected } errors, and fixing the CI.

I am working on it either, I get the same `Length Required` errror and `TrySendError { kind: Disconnected }` when I run test with `cargo test --features=cli,env-file`.

ramsayleung commented:
I am trying to update Github Action CI configuration to run `cargo clippy` separately for feature `client-ureq` and `client-reqwest`, but turns out I fail to build `rspotify` with features `client-ureq`:

```sh
cargo build --features client-ureq
```
```log
error[E0592]: duplicate definitions with name `from_response`
  --> src/http/reqwest.rs:14:5
   |
14 |     pub async fn from_response(response: reqwest::Response) -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definitions for `from_response`
   |
  ::: src/http/ureq.rs:11:5
   |
11 |     pub fn from_response(r: ureq::Response) -> Self {
   |     ----------------------------------------------- other definition for `from_response`

error[E0592]: duplicate definitions with name `request`
  --> src/http/reqwest.rs:50:5
   |
50 | /     async fn request<'a>(
51 | |         &self,
52 | |         method: Method,
53 | |         url: &str,
54 | |         headers: Option<&Headers>,
55 | |         payload: Option<Content<'_>>,
56 | |     ) -> ClientResult<String> {
   | |_____________________________^ duplicate definitions for `request`
   |
  ::: src/http/ureq.rs:17:5
   |
17 | /     fn request<'a>(
18 | |         &self,
19 | |         req: &mut ureq::Request,
20 | |         headers: Option<&Headers>,
21 | |         payload: Option<Content<'a>>,
22 | |     ) -> ClientResult<String> {
   | |_____________________________- other definition for `request`


```

marioortizmanero commented:
You have to use `cargo build --no-default-features --features=env-file,cli,client-ureq` because you have to disable the default client as well, `client-reqwest`. I wanted to add a note for that in the docs.

marioortizmanero commented:
I've added proper `cargo clippy` to the CI and fixed a couple things. As you can see, we get warnings for the following:

```
error: unnecessary parentheses around block return value
   --> src/client.rs:944:19
    |
944 |         self.me().await
    |                   ^^^^^ help: remove these parentheses
```

but these are a bug in `maybe_async` that I wanted to report.

marioortizmanero commented:
I've also updated the documentation. Give it a read and let me know what you think!

ramsayleung commented:
>  You have to use cargo build --no-default-features --features=env-file,cli,client-ureq because you have to disable the default client as well, client-reqwest. I wanted to add a note for that in the docs.

And I have also updated the documentation :).

There is one more thing I want to discuss, is it necessary to introduce `ureq` to our dependencies? I know it's more lightweight that `reqwest` in some way, but introduce `ureq` will break the consistency of HTTP client, as we don't need two or more HTTP clients, all we want is just a way to switch between `async` and `blocking`, IMHO, `ureq` is not that necessary, one client is enough.

marioortizmanero commented on src/lib.rs:
```diff
@@ -197,3 +242,5 @@ mod macros {
         };
     }
 }
+//!  (and
```
This is out of place

marioortizmanero commented on src/lib.rs:
```diff
@@ -161,20 +201,25 @@
 //! and blocking client, which is triggered inside `Cargo.toml`, so there is
 //! something you need to pay attention to when you are trying to build `rspotify`
+//!
+//! Build with `client-reqwest` feature, the `async` version, and this would compile Rspotify with [`reqwest`](https://docs.rs/reqwest/)
```
AFAIK these build instructions are only for `rspotify` itself. Are they meant for developers of this library? In that case I think they'd fit better in the README, as it's only hosted on GitHub

marioortizmanero commented:
> There is one more thing I want to discuss, is it necessary to introduce ureq to our dependencies? I know it's more lightweight than reqwest in some way, but introducing ureq will break the consistency of HTTP client, as we don't need two or more HTTP clients, all we want is just a way to switch between async and blocking, IMHO, ureq is not that necessary, one client is enough.

I'd say so, yeah. I think we differ in opinions: you just want to switch between sync and async, and I want to switch between multiple HTTP clients, which comes with sync/async switching as kind of a side effect. The point of this PR is to be able to use multiple clients at the same time. `ureq` is just one of them, but anyone in the future could add support for another popular HTTP client that rises up; this solution is the most flexible one I could come up with. I'm pretty sure if you asked any `rspotify` user which one they prefer, they'd go with multiple clients. But is of course made up, we could ask more people about their opinion.

As to `ureq`, it's a much more lightweight client than `reqwest::blocking`, and designed to be synchronous, so those using this library with blocking code won't have to go through `reqwest`'s overhead. This requires 95 compilation units:

```
cargo build --no-default-features --features=client-ureq,ureq-rustls-tls
```

And this one is *152*:

```
cargo build --no-default-features --features=client-reqwest,reqwest-rustls-tls
```

It might make development a bit more complicated because we don't have some features from `reqwest`, but I think these are minimal and can be worked around.

ramsayleung commented:
> I've fixed the remaining errors. I was getting a Length Required one because POST, PUT and DELETE methods can't have empty payloads. I'm still trying to figure out why I get some TrySendError { kind: Disconnected } errors, and fixing the CI.

As for `TrySendError { kind: Disconnected }` I am still struggling with it, but my intuition is there is something wrong in `async_once`, probably race condition bug. Because when I change `CREDS_TOKEN.get().await` to `get_creds_token().await`, all tests are passed, except for `test_fake_playlist`, which is nothing related to `TrySendError { kind: Disconnected }`. And when I run all tests simultaneously, I get `TrySendError { kind: Disconnected }`, but when I run test one by one, this error just disappeared, so probably something wrong with `async_once`.

![image](https://user-images.githubusercontent.com/14180681/94575135-7f63b680-02a6-11eb-8eaf-733fb5ecc178.png)

marioortizmanero commented:
Not sure what the fake playlist request is for anyway? Why use a fake id?

marioortizmanero commented:
Ok so I've properly fixed the tests. See the commit message for https://github.com/ramsayleung/rspotify/pull/129/commits/6adde825714ea8a8dc4d321782e9c62c840121d5.

ramsayleung commented:
> Not sure what the fake playlist request is for anyway? Why use a fake id?

Just using a fake id for exceptional test case, to check whether could `rspotify` handle it properly or not.

marioortizmanero commented:
> Just using a fake id for exceptional test case, to check whether could rspotify handle it properly or not.

Ah ok we'll leave it then.

> Same here, maybe it is caused by ureq/tls which doesn't support cross-compile?

I tried to run the tests without ureq and they aren't working either, so it's not that: https://github.com/ramsayleung/rspotify/runs/1192215892

marioortizmanero commented:
I've found the problem in the CI. I wanted to reduce the number of jobs when testing and building, because we now have to compile *for each architecture* and *for each client*, which may blow up the number of jobs when more are added. But it's dumb to try to test when cross-compiling because you aren't actually on a machine on that architecture. For now, I think we can keep it as it was before, with building and testing in separate jobs.

marioortizmanero commented:
Believe it or not, but tests are actually passing! You can take a look at the code when you have time now, as I won't be making any more changes. No rush either, it's a ton of code. I will be creating new issues in the upstream libraries that use rspotify for opinions on breaking changes we can make. I also wanted to analyze the changes: how many dependencies we removed, compile & runtime performance... But I might wait for pre-release before that.

marioortizmanero commented:
Ah, I had forgotten that the `webapp` example has to be updated. Do you want to do that, @ramsayleung, or shall I take care of it?

ramsayleung commented:
> Do you want to do that, @ramsayleung, or shall I take care of it? 
I will update the `webapp` example :)

marioortizmanero commented:
I think that's because you're missing the TLS implementation in the `features`. If you want to use `reqwest` with the default TLS you can just do this:

```
RUST_LOG=info cargo run --example oauth_tokens --features env-file,cli
```

And if you wanted to change the client or the TLS, you'd have to specify both of them:

```
RUST_LOG=info cargo run --example oauth_tokens --features env-file,cli,client-reqwest,reqwest-rustls-tls --no-default-features
```

or

```
RUST_LOG=info cargo run --example oauth_tokens --features env-file,cli,client-ureq,ureq-rustls-tls --no-default-features
```

marioortizmanero commented:
> By the way, I have change get_token, get_creds and get_oauth functions to be globally public instead of (in crate) public. Because I think it makes sense to check whether token exists before to request for it, which is a common use case in Web App.

No, you don't need to check `get_creds` or `get_auth` explicitly. The functions that need them will use them internally and return an `InvalidAuth` error, which is what you need to check for instead.

ramsayleung commented:
> No, you don't need to check get_creds or get_auth explicitly.

Yes, it makes sense, I will revert these two functions.

marioortizmanero commented:
I've edited my comment with an example of how you could make it work btw

ramsayleung commented:
By the way, will you get same error as I do when you run example `oauth_tokens`?

marioortizmanero commented:
I think you missed this comment https://github.com/ramsayleung/rspotify/pull/129#issuecomment-706674025

ramsayleung commented:
Sorry for missing your comment, I have fixed the `webapp` example :)

marioortizmanero commented:
Ok so I've made a few discussion-related issues recently. The plan is to open some more issues in libraries that use rspotify like spotifyd, spotify-tui or ncspot, and both let them know about the rewrite and ask for their opinion. That should help us to take the decisions. I'll do that once this PR is merged.

ramsayleung commented:
I have finished the code review, it's a huge PR, takes many time to review, and I have added a few suggestions, you could take time to have a look at them :)

marioortizmanero commented:
Sorry what suggestions?

ramsayleung commented:
I have left some review comments, does you miss them? And I have merged a PR submitted by other contributor which seems to cause a conflict on file `CHANGELOG.md`

marioortizmanero commented:
I'm afraid I can't find these comments. Can you provide a link? I'll fix the CHANGELOG.md issue.

ramsayleung reviewed (commented):
There are all my review comments, I just forgot to submit my reviews and all comments are pending, sorry for that.

marioortizmanero commented on src/client.rs:
```diff
@@ -100,140 +95,153 @@ pub enum ApiError {
+    /// The cache file path, in case it's used. By default it's
+    /// `.spotify_token_cache.json`.
+    #[builder(default = "PathBuf::from(\".spotify_token_cache.json\")")]
+    pub cache_path: PathBuf,
```
Checking if the cache exists is already an unnecessary overhead, I would keep it this way to maintain `get_token_without_cache` actually cache-free. If anything, that could be discussed in #135.

marioortizmanero commented on examples/album.rs:
```diff
@@ -1,22 +1,41 @@
+    // token will be modified. We don't need OAuth for this specific endpoint,
+    // so `...` is used instead of `prompt_for_user_token`.
+    spotify.request_client_token().await.unwrap();
+
```
I wouldn't do that implicitly. The client token is only for the Client Credentials token, so it's unnecessary for other methods. That way you can construct a `Spotify` type without using `await` or just blocking. The cache file is already read before `request_client_token` actually requests a token AFAIK.

marioortizmanero commented on src/oauth2.rs:
```diff
@@ -172,35 +168,19 @@ impl Spotify {
 
         if show_dialog {
-            json_insert!(payload, "show_dialog", "true");
+            payload.insert("show_dialog", "true");
```
Not sure what you mean. `reqwest` is one of the HTTP clients we support, we can't rely on it. I'm using the `url` crate for parsing, which most clients use underneath anyway so it's not much of a problem.

ramsayleung replied:
Yep, I get it. My intutuion is that we could use `reqwest` to handle url parse, but turns out that `reqwest` isn't enable all the time, using `reqwest` to parse url is not a good idea.

marioortizmanero commented on src/http/ureq.rs:
```diff
@@ -0,0 +1,108 @@
+        send_request: D,
+    ) -> ClientResult<String>
+    where
+        D: Fn(&mut Request) -> Response,
```
It's just a function to continue building the `Request` like the one in `src/http/reqwest.rs`. You can see its usage later on. In this case, calling something like `send_json` with `ureq` converts the Request into a Response, so that's the function signature needed. Using a closure makes it straightforward to choose between GET/POST/PUT/DELETE, and avoids the `Content::Json` and `Content::Form` enum I had previously, which was admittedly a mess.

marioortizmanero commented on src/oauth2.rs:
```diff
@@ -1,576 +1,420 @@
+    #[maybe_async]
+    pub async fn request_client_token_without_cache(&mut self) -> ClientResult<()> {
+        let mut data = Form::new();
+        data.insert("grant_type".to_owned(), "client_credentials".to_owned());
```
What is hardcoded? The `grant_type`? Sure, that could be added to the `http` module.

marioortizmanero replied:
done

marioortizmanero commented on src/client.rs:
```diff
@@ -403,166 +431,184 @@ impl Spotify {
+        params.insert("offset".to_owned(), offset.into().unwrap_or(0).to_string());
+        params.insert("q".to_owned(), q.to_owned());
+        params.insert("type".to_owned(), _type.as_str().to_owned());
+        if let Some(market) = market {
```
I did think about it. I initially wanted to use a `serde_json::Value` type for GET requests too because it makes its usage super easy and clean, but I ended up running into too many issues.

I also tried to make the query type a `HashMap<&str, &str>` but failed because of ownership issues (this would avoid the `to_owned` calls). All of the following conditional inserts failed because `&str` is not owned and it goes out of scope:

```rust
        if let Some(market) = market {
            params.insert("market", market.as_str());
        }
```

But I can give that a try again to see if it's possible to improve before being forced to use a macro (you can give it a try as well). Rather than creating one I'd use existing crates, like [maplit](https://github.com/bluss/maplit).

marioortizmanero replied:
I would leave this for a different PR, though.

marioortizmanero commented on tests/test_with_credential.rs:
```diff
@@ -1,166 +1,177 @@
-}
+/// Generating a new basic client for the requests.
+#[maybe_async]
+pub async fn creds_client() -> Spotify {
```
No. I initially wanted to reuse the same client, but ran into lots of different issues related to concurrency or `lazy_static` (not sure if you remember when I was using `async_once`, but this had race conditions for some reason). I then undid my changes to use different clients, because tests should be ran in an "isolated" environment anyway.

marioortizmanero commented on src/oauth2.rs:
```diff
@@ -1,576 +1,420 @@
+            .unwrap();
 
-    #[test]
-    fn test_parse_response_code() {
```
I wanted to get rid of the `parse_response_code` and just left a TODO for later. But it's relatively easy so I've just made a commit with a solution that is guaranteed to work thanks to the `url` crate. 

The function signature should actually be `ClientResult<Option<String>>` because it may return `ParseError` when the provided URL is wrong, and `None` when the `code` parameter isn't found, but I thought this would be less annoying to deal with. Not sure what you think.

ramsayleung replied:
> I wanted to improve parse_response_code's implementation to use an external crate and just left a TODO for later.

I think it's not that complex to parse the response code so that we don't need an external crate to parse code. And we still keep the function `parse_response_code`, we should remain its test either, or removing it util introducing an external crate.

> The function signature should actually be ClientResult<Option<String>> because it may return ParseError when the provided URL is wrong, and None when the code parameter isn't found, but I thought this would be less annoying to deal with

Great idea, it makes sense.

marioortizmanero replied:
> I think it's not that complex to parse the response code so that we don't need an external crate to parse code. 

Yes, it's that complex. The previous solution was incorrect. If the code parameter wasn't the first one it wouldn't work at all, for example. We already have the `url` crate for other stuff like `get_authorize_url` so there's no problem using it.

> And we still keep the function parse_response_code, we should remain its test either, or removing it util introducing an

Sure

ramsayleung commented on src/oauth2.rs:
```diff
@@ -1,576 +1,416 @@
+        data.insert("redirect_uri".to_owned(), oauth.redirect_uri.clone());
+        data.insert("code".to_owned(), code.to_owned());
+        data.insert("scope".to_owned(), oauth.scope.clone());
+        data.insert("state".to_owned(), oauth.state.clone());
```
Perhaps we should replace these hardcoded string `grant_type`, `redirect_uri `, `code` with constant variable or enum either.

marioortizmanero replied:
done

ramsayleung commented on src/http/mod.rs:
```diff
@@ -0,0 +1,116 @@
+    pub const RESPONSE_CODE: &str = "code";
+    pub const RESPONSE_TYPE: &str = "response_type";
+    pub const SCOPE: &str = "scope";
+    pub const SHOW_DIALOG: &str = "state";
```
Is it a problem? SHOW_DIALOG should be `show_dialog`?

marioortizmanero replied:
Whoops! Fixed.

ramsayleung commented:
Is it ready to merge? :)

marioortizmanero commented:
I would say so!

ramsayleung commented:
Merged :)
